### ISSUE TITLE
research: explore a learned lightweight coordinator (Trinity-style role-assignment loop) for Looper-family routing

### ISSUE THREAD

**Issue by Peterren** [vllm-project/semantic-router#2383]
## Summary

Explore a new Looper-family routing algorithm in which a **small, learned coordinator** decides, turn by turn, **which model to call and what role to assign it** (plan / execute / verify), instead of using a fixed deliberation topology. The coordinator is a compact model plus a tiny classification head; it is trained offline against a terminal reward and never needs to be as capable as the models it orchestrates.

This is an exploration/design issue, not a request to ship a production algorithm. The goal is a prototype and an evaluation write-up that can be compared against existing Looper, Fusion, and ReMoM paths under the shared benchmark protocol.

Parent roadmap: #2287
Parent issue: #2337 (explore additional Looper-family routing algorithms)
Depends on: #2333 (decision-level eval harness), #2336 (harden existing Looper/Fusion/ReMoM)
Related: #2346 (compare Router Learning algorithms), #2371 (Looper-family benchmark suite), #1815 (multi-step model pipelines), #2308 (contextual bandit strategies)
Related code: `src/semantic-router/pkg/looper` (esp. `rl_driven.go`, `workflows_planner.go`, `remom.go`, `confidence.go`), `src/semantic-router/pkg/selection`, `bench`

## Motivation

Semantic Router already has multi-model deliberation (Looper, Fusion, ReMoM) and one research-derived RL loop (`rl_driven.go`, integrating Router-R1, arXiv:2506.09033). These are valuable, but the *coordination policy* — who to call, in what order, and to do what — is largely hand-designed or confidence/rating driven.

A complementary direction is to **learn the coordination policy itself** with a lightweight model, so that per-query orchestration (topology + role assignment) emerges from data instead of being prescribed. Recent work shows this is feasible with a very small coordinator and modest training budget, which fits the router's "add capability through clear contracts, not ad hoc special cases" principle.

## Inspiration

- **TRINITY: An Evolved LLM Coordinator** — Sakana AI, ICLR 2026 (arXiv:2512.04695). A ~0.6B SLM + ~10K-param head coordinates a pool of frozen LLMs over multi-turn interaction. Each turn it selects one LLM and assigns it a role — **Thinker** (strategize/decompose), **Worker** (execute), or **Verifier** (accept/revise). It halts when a Verifier accepts, or at a fixed turn budget. Crucially, the coordinator reads the SLM's **hidden state**, not its generated text, and is trained with **sep-CMA-ES** (a derivative-free evolutionary strategy), avoiding backprop through any LLM.
- **Learning to Orchestrate Agents with the Conductor** — Sakana AI, ICLR 2026 (arXiv:2512.04388). A 7B RL-trained coordinator that designs communication topologies and prompts across a randomized worker pool.
- **Sakana Fugu** (product form of both): https://sakana.ai/fugu-beta/
- Precedent in this repo: `looper/rl_driven.go` already integrates a routing research paper (Router-R1) into the Looper framework, so a Trinity-style coordinator would follow an established integration pattern.

## Goals

- Define a `Coordinator` Looper algorithm contract: at each turn, given the transcript, choose `(model, role)` where `role ∈ {thinker, worker, verifier}`, with a fixed turn budget K and a Verifier-accept stop condition.
- Prototype the coordination loop **first with a rule/heuristic coordinator** (no training) so the loop, role prompts, transcript handling, stop rule, and diagnostics can be validated cheaply and deterministically.
- Add a **learned coordinator** variant (compact model + small head) trained offline against a terminal reward; keep the entire worker pool frozen. Start with an open-source worker pool for reproducible, low-cost evaluation.
- Make the algorithm measurable through the decision-level eval harness (#2333) and the Looper benchmark suite (#2371): report quality, cost/token overhead, latency, turn count, and per-role call distribution.
- Emit replay/diagnostics that explain, per turn: selected model, assigned role, stop reason, and budget consumption.

## Non-goals

- Do not ship a production algorithm in this issue; productionization is a follow-up gated by eval + protection coverage.
- Do not turn Semantic Router into a general workflow engine — this stays inside the Looper-family contract with strict turn/model/token caps.
- Do not require frontier/proprietary APIs for baseline CI; open-source workers must be sufficient to reproduce the mechanism.
- Do not train or bundle any model weights into the router repo; training and any coordinator checkpoint live outside the request path.
- Do not reproduce leaderboard SOTA numbers. The aim is the *mechanism* (learned role-assignment coordination beating the best single worker under a matched budget), not matching the paper's absolute scores.

## Proposed direction (high level)

Keep the router's pipeline intact and add this as a Looper algorithm behind the existing contract:

    request -> signals -> decision -> model_selection
                                          |
                                          v
                         Looper: Coordinator loop (turns 1..K)
                          each turn: coordinator picks (model, role)
                          role-specific prompt -> selected worker LLM
                          append result to transcript
                          stop when Verifier accepts or budget K hit

- Reuse `looper` transcript/state, model-ref caps, token filters, and SSE/usage plumbing rather than adding a parallel execution path.
- The coordination policy is pluggable: a deterministic heuristic coordinator for CI, and an optional learned coordinator for research runs.
- Training (offline, out-of-tree): terminal-reward optimization of the small head (+ optional singular-value fine-tuning of a few backbone layers). Evolutionary strategy (sep-CMA-ES) is the primary candidate per Trinity; this is compared, not assumed, against an RL baseline.

## What we want to achieve (success picture)

- A working `Coordinator` Looper algorithm with a deterministic mode that passes CPU CI and a documented learned mode.
- A benchmark report on a small, cheap task subset (e.g. MATH500 test split) showing the learned coordinator, using an **open-source worker pool**, beats the best single worker in that pool under a matched output-token budget.
- Clear cost/latency/turn-count accounting and replay diagnostics, so reviewers can see *why* a route was chosen.
- A design note comparing evolutionary vs RL training for this coordination objective, so a future productionization decision has evidence.

## Acceptance criteria

- The `Coordinator` algorithm contract (turns, roles, model caps, token caps, stop rule, failure/timeout behavior) is documented.
- A deterministic-coordinator prototype runs inside the Looper framework and is covered by CPU-only tests.
- At least one learned-coordinator experiment is evaluated against single-worker and existing Looper baselines through the decision-level eval harness (#2333).
- Results include quality, cost, token overhead, latency, and per-role/turn diagnostics with regression thresholds.
- Productionization is deferred to explicit follow-up issues; no exploratory training code lands silently in the request path.

## Validation (planned)

    # deterministic coordinator, CPU-only
    make test  # add coordinator loop unit tests under pkg/looper

    # research eval (open-source worker pool, small subset)
    # coordinator vs best-single-worker under matched token budget
    # report via decision-level eval harness (#2333) + looper bench (#2371)

**Comment by Peterren** [comment]
## Exploration update: a Trinity-style learned coordinator for per-query model+role routing

**TL;DR** — I built a minimal reproduction of the *mechanism* of **Trinity** (Sakana AI, arXiv:2512.04695) as a Looper-family routing strategy and evaluated it honestly on GPQA-Diamond. Three takeaways:
1. **It works, scoped:** on a *balanced, complementary* 3-model pool, a tiny learned coordinator beats every baseline significantly on held-out GPQA (0.652 vs best-single 0.529, round-robin 0.558; p<0.05 vs each).
2. **But it does not transfer:** the same policy, applied to a different domain without retraining, collapses back toward always-best-single (not significant). Learned routing is domain-specific.
3. **The reusable lesson for Looper:** *pool structure decides whether routing can help at all*, and a learned router needs to be per-recipe, cheap to retrain, and drift-guarded. Concrete asks at the bottom.

Not proposing to productionize — this is the "prototype + evaluation write-up" this issue asked for.

### Setup (compressed)
- **Coordinator**: frozen **Qwen3-0.6B** → take its penultimate-token **hidden state** (text discarded) → tiny **~6K-param linear head** → outputs per turn *(worker model, role in {Thinker, Worker, Verifier})*.
- **Training**: **sep-CMA-ES** (derivative-free, no backprop through any LLM), fitness = mean binary terminal reward over a fixed 60-question train subset, common-random-numbers for paired comparison. K≤2 turns; stop on Verifier ACCEPT; per-turn thinking cap with overflow logged (never silently scored wrong).
- **Workers (frozen, vLLM)**: **Qwen3-32B** (reasoning) + **Mistral-Small-3.2-24B** (non-reasoning) + **phi-4** (synthetic-data) — chosen via a complementarity pre-check for balanced strength + orthogonal errors.
- **Task**: GPQA-Diamond, fixed 60-train / 138-held-out split. Reward = final boxed letter == gold.

### Results

**Main (held-out GPQA-D, n=138, K=2):**

| system | resolve | 95% CI |
|---|---|---|
| single Qwen3-32B / Mistral / phi-4 | 0.493 / 0.471 / **0.529** | |
| random coordinator | 0.543 | 0.46-0.62 |
| round-robin (T→W→V) | 0.558 | 0.47-0.64 |
| **learned coordinator** | **0.652** | **0.570-0.727** |
| oracle (per-query best-of-3) | 0.850 | |

McNemar (paired) beats **every** baseline: vs round-robin p=0.043, vs best-single p=0.010, vs random p=0.029, vs singles p=0.0015. Routes **non-trivially** (worker mix phi-4 140 / Qwen3 117 / Mistral 19 — not always-best-single). Seed-robust: 3 CMA seeds → 0.652 / 0.623 / 0.638 (sd 0.012). Captures ~38% of the oracle headroom (best-single 0.529 → oracle 0.850).

**Ablation (K=1 vs K=2):** K=1 (one turn) → **0.594** (already beats all baselines); K=2 (up to two turns) → **0.652**. Important caveat: the learned policy almost always picks the **Verifier** role (88% of turns; Thinker never chosen) and never stops early, so the K=2 gain comes from a **second attempt / second model**, **not** from an exploited Thinker→Worker→Verifier division of labor. In fact the role dimension collapsed — my role prompts weren't mutually exclusive (both Worker and Verifier were asked to emit the final `\boxed{}` answer, and Verifier could answer with no prior solution to check), so evolution just kept the one role that answers in a single step. Net: the win is from **per-query model selection + a second try**, and the *role structure as implemented was not the lever*.

**Cross-domain transfer (the key negative):** the GPQA-trained coordinator, run unchanged on **MMLU humanities** (law/religion/ethics, n=120, same 4-option format) beats best-single by only **+0.8 pt (p=1.000, n.s.)** and its worker mix **collapses toward phi-4** (179/29/21), vs the balanced 140/117/19 in-domain. The routing policy learned on science **does not transfer** — it degrades toward always-best-single. Retraining is cheap (~2 min of CMA once worker outputs are cached), but a new distribution needs it. (This matches Trinity's fixed-pool design; cross-domain generality is what the sister **Conductor** paper, arXiv:2512.04388, targets via RL + randomized-agent-pool training — a heavier regime.)

### Implications for the Looper family (the actionable part)
Concrete, repo-relevant suggestions any multi-model Looper algorithm (Fusion, ReMoM, a future coordinator) could adopt:

1. **Gate on pool complementarity before routing.** Routing only pays off when the `modelRefs` pool is balanced (no dominator) and complementary (orthogonal errors). We hit both failure modes: two similar reasoning models → overfit, *below random*; adding one dominating model → degenerate "always pick the strong one". A cheap pre-check (per-model accuracy + pairwise both-correct + **union-oracle vs best-single**) tells you if a `modelRefs` set has *any* headroom. If union-oracle ≈ best-single, fall back to best-single instead of spending N× tokens for nothing. Good fit for Looper config validation.
2. **Make round-robin a mandatory baseline.** On a complementary pool a *dumb* round-robin (0.558) beat every single model — a learned/complex algorithm must beat *round-robin*, not just best-single, to justify its cost.
3. **Report value as fraction of oracle headroom, not raw delta.** 0.652 looks modest until you see it's ~38% of the reachable gain. A Looper replay/dashboard showing **oracle-vs-achieved** is far more informative than raw accuracy.
4. **Confidence-gate deliberation depth — and make role contracts mutually exclusive.** ~half the gain is per-query model-pick (K=1); a second turn adds the rest, while cost grows ~(models×roles)^K — so start shallow and escalate only on low confidence (aligns with the existing `confidence` path). Caveat from our run: the K=2 gain was a *second attempt/second model*, NOT an exploited role division — our policy collapsed to always-Verifier because the Thinker/Worker/Verifier prompt contracts overlapped (multiple roles could emit the final answer). Lesson for a role-based Looper algorithm: enforce **mutually-exclusive role contracts** (only Worker emits an answer; Verifier only judges ACCEPT/REVISE on an existing answer and cannot answer from scratch), or the optimizer will collapse to whichever role answers in one step and the role structure buys nothing.
5. **A learned selector must be per-recipe + cheap to retrain + drift-guarded.** Because it *silently* degrades under distribution shift (no error, just worse), train it per-decision/per-recipe (fits this repo's decision→modelRefs structure), keep it lightweight (~6K params, ~2 min retrain), and monitor routing-distribution drift with a fallback to round-robin/best-single.

### Scope / limitations
Single benchmark (GPQA-D), n=138 held-out, one split; K≤2; head-only (deferred the paper's singular-value fine-tuning); 3-model pool; binary reward; one domain-shift probe (n=120). Exploration result, not a production claim. Happy to adapt the eval harness or coordinator to this repo's conventions if maintainers think it's worth a follow-up.

**Comment by Xunzhuo** [comment]
Maintainer consolidation: the exploration delivered a substantive evaluation and actionable conclusions, so this research issue is complete. The results—pool complementarity, matched-budget baselines, per-recipe retraining, drift guards, and the role-collapse caveat—should inform the accepted long-session and Looper initiative #2973 under #2965. Production or partner work will be scoped there rather than keeping a second roadmap item open.
